### ISSUE TITLE
Non-bootstrapped ForkedJvmInfo causes NullPointerExceptions

### ISSUE THREAD

**Issue by softwaredoug** [randomizedtesting/randomizedtesting#250]
Hello! Perhaps more of a question than an issue. Building Elasticsearch on OS X, I noticed [this issue](https://discuss.elastic.co/t/periodic-nullpointerexceptions-in-heartbeatevent-running-gradle-test-in-os-x/91014) that was driving me crazy, thinking I had a bug in my code :)

In short, I get periodic stack traces:

```
Exception thrown by subscriber method onHeartbeat(com.carrotsearch.ant.tasks.junit4.events.aggregated.HeartBeatEvent) on subscriber com.carrotsearch.gradle.junit4.TestReportLogger@52ca2fa9 when dispatching event: com.carrotsearch.ant.tasks.junit4.events.aggregated.HeartBeatEvent@3fa57c62
java.lang.NullPointerException
        at com.carrotsearch.ant.tasks.junit4.ForkedJvmInfo.getPidString(ForkedJvmInfo.java:85)
        at com.carrotsearch.ant.tasks.junit4.ForkedJvmInfo$getPidString.call(Unknown Source)
        at com.carrotsearch.gradle.junit4.TestReportLogger.onHeartbeat(TestReportLogger.groovy:116)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
```

I've been digging into this, and it seems to perhaps be a randomized testing issue, related to [this log statement](https://github.com/randomizedtesting/gradle-randomized-testing-plugin/blob/master/src/main/groovy/com/carrotsearch/gradle/junit4/TestReportLogger.groovy#L116).

Which ultimately invokes [ForkedJvmInfo::getPidString](https://github.com/randomizedtesting/randomizedtesting/blob/master/junit4-ant/src/main/java/com/carrotsearch/ant/tasks/junit4/ForkedJvmInfo.java#L84) which doesn't check if bootstrapEvent has been set:

```
  public String getPidString() {
    return bootstrapEvent.getPidString();
  }
```

Which is odd, because other parts of ForkedJvmIssue appear to do such null checks, such as [here](https://github.com/randomizedtesting/randomizedtesting/blob/master/junit4-ant/src/main/java/com/carrotsearch/ant/tasks/junit4/ForkedJvmInfo.java#L66).

Further digging lead me to presuppose bootstrap and heartbeat might be independent events, so perhaps heartbeat occurs before bootstrap. And if that happened, the log statement would throw a NullPointerException.

I'm lead to believe either:
- I should be concerned: Something bad happened, and bootstrap event should have been set, and I'm actually in a bad state
- I don't need to worry (my working assumption): that sometimes a heartbeat would fire randomly before a bootstrap event. So I can safely avoid this NullPointerException and this is a minor bug.

I'm using MacOSX, and have seen this with Gradle 3.5/4.0 across two versions of Oracle JDK 8.

**Comment by dweiss** [comment]
The bootstrap event is sent back from the forked JVM immediately after it starts... it's very strange that heartbeat is invoked before this happens.

https://github.com/randomizedtesting/randomizedtesting/blob/master/junit4-ant/src/main/java/com/carrotsearch/ant/tasks/junit4/slave/SlaveMain.java#L336-L339

So while patching the getPidString method may be a workable fix, it doesn't reveal the cause why this race condition happens in the first place -- it shouldn't, really. What is your heartbeat interval? Could it happen that it's reached before all the JVMs boot up properly (heavily loaded system or I/O queues)?

**Comment by softwaredoug** [comment]
I'm running Elasticsearch's tests. Heartbeat interval looks like its set to 10 (is that seconds?).

Is there a way I can attach a debugger to the forked JVMs? Such as by passing an environment variable or some such to pass the right debugger options?

Could something be happening that causes the Forked JVM to stall before getting to bootstrap event?

**Comment by dweiss** [comment]
I didn't write the groovy runner (Ryan Ernst did), but I think it uses the Ant code underneath, so it should be the same... There is no need to attach to the forked JVM - it'd be enough if you could just capture more verbose ant runner logs (this is forked from groovy). You'll probably need to change the runner's code here:

https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/groovy/com/carrotsearch/gradle/junit4/RandomizedTestingTask.groovy#L214-L234

Once you have those logs, they should include messages captured from the forked JVMs -- these should be more informative as to what is actually happening. If you can make these logs available, I'll take a look. Oh, thanks for doing this by the way!

**Comment by s1monw** [comment]
@rjernst what do you think about @dweiss suggestions here ^^

**Comment by softwaredoug** [comment]
Thanks! I ran Elasticsearch's `integTest` task. Attached zipped up log.

It does look like the heartbeat comes in before the bootstrap event. The sequence of events seems to be

JVM's starting up

```
21:00:00.157 [DEBUG] [org.gradle.api.internal.project.ant.AntLoggingAdapter] [ant:junit4] Starting JVM J0
21:00:00.158 [DEBUG] [org.gradle.api.internal.project.ant.AntLoggingAdapter] [ant:junit4] Starting JVM J1
21:00:00.161 [DEBUG] [org.gradle.api.internal.project.ant.AntLoggingAdapter] [ant:junit4] Starting JVM J2
```

Then Heartbeat event & NPE:

```
21:00:10.205 [ERROR] [com.carrotsearch.ant.tasks.junit4.dependencies.com.google.common.eventbus.EventBus.aggregated] Exception thrown by subscriber method onHeartbeat(com.carrotsearch.ant.tasks.junit4.events.aggregated.HeartBeatEvent) on subscriber com.carrotsearch.gradle.junit4.TestReportLogger@6d4c57c2 when dispatching event: com.carrotsearch.ant.tasks.junit4.events.aggregated.HeartBeatEvent@5e778f61
java.lang.NullPointerException
```

Then finally bootstrapping

```
21:00:10.752 [INFO] [org.gradle.api.Task] Started J2 PID(45341@wiz.local).
21:00:10.752 [DEBUG] [org.gradle.api.internal.project.ant.AntLoggingAdapter] [ant:junit4] Packet received, slave#2>BOOTSTRAP
21:00:10.754 [DEBUG] [org.gradle.api.internal.project.ant.AntLoggingAdapter] [ant:junit4] Packet received, slave#2>SUITE_STARTED
```

[integTest2.log.zip](https://github.com/randomizedtesting/randomizedtesting/files/1116776/integTest2.log.zip)

**Comment by softwaredoug** [comment]
@dweiss you pointed out that bootstrap should be emitted immediately. Noticeably, there's several synchronize statements around a lock [in the `Serializer`](https://github.com/randomizedtesting/randomizedtesting/blob/master/junit4-ant/src/main/java/com/carrotsearch/ant/tasks/junit4/events/Serializer.java#L84) used to post bootstrap. If that the JVM is blocked on that lop, then bootstrap might not be emitted immediately--letting other events like Heartbeat sneak in.

**Comment by dweiss** [comment]
Thanks Doug. I'll take a look later today. Somehow I don't think it's a serialization/synchronization problem -- parly because the heartbeat is currently "faked" by events on the master side only. Looks like the load on starting those JVMs takes longer than I ever expected it could, causing a race on the master. Ideally, the heartbeat should be emitted from the forked JVM (there were reasons why this wasn't implemented this way, but I'll omit them here). 

I'll take a look and try to reproduce it, then I'll fix it. It's a bug.

**Comment by dweiss** [comment]
Ok, I can reproduce it. Will fix it soon, thanks Doug.

**Comment by dweiss** [comment]
This commit should fix it, Doug. I'll try to release shortly; if you can try it out locally, it'd be great. Separately from this fix, the heartbeat should really be based on forked JVM events (and possibly the master monitoring the forked JVM state, just in case the JVM hangs). I'll get down to it at some point...

**Comment by softwaredoug** [comment]
Thanks for the help!

**Comment by dweiss** [comment]
Released 2.5.2, should be ok now.

**Comment by s1monw** [comment]
I opened a PR to upgrade ES to this. Thanks for sorting this out @softwaredoug @dweiss
